File: 70. Climbing Stairs.java
public class Solution {
    public int climbStairs(int n) {
        Map<Integer, Integer> hs = new HashMap<Integer, Integer>();
        int output = count(n, hs);
        return output;
    }
    
    public int count(int curr, Map<Integer, Integer> hs) {
        int val = 0;
        
        if (hs.containsKey(curr)) {
            return hs.get(curr);
        } else {
            if (curr < 1) {
                val = 0;
            } else if (curr == 2) {
                val = 2;
            } else if (curr == 1) {
                val = 1;
            } else {
                val = count(curr - 1, hs) + count(curr - 2, hs);
            }
            hs.put(curr, val);
        }
        return val;
    }
}

File: 70. Climbing Stairs.py
class Solution:
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        def climb(n, curr, memo):
            count = 0
            
            if curr in memo:
                return memo[curr]
            if curr == n:
                return 1
            elif curr > n:
                return 0
            
            count += climb(n, curr + 1, memo) + climb(n, curr + 2, memo)
            memo[curr] = count
            
            return count
        
        return climb(n, 0, {})
